### ISSUE TITLE
Propel InstancePooling must be disabled for querying tables without a PrimaryKey (such as Views)

### ISSUE THREAD

**Issue by j5bart** [propelorm/Propel2#797]
Firstly, thanks to marcj for helping me work around my issue on gitter! 

This is how the conversation went:
# 

killercowuk
34 minutes ago
Hi all, relatively new to Propel - I'm setup and starting to run queries, but for read-only/views that I've added to schema and mapped, I'm having a strange issue when running Queries - I get a results returned but they all have the same data, like the first row is repeated. The result COUNT is correct though. Has anyone experienced this before?
I've logged the SQL and ran it manually - the database returns the expected results. It seems to be somewhere in ObjectFormatter that the data of rows after the first is not being replaced with the next rows content or something...
any ideas/recommendations greatfully received.
marcj
25 minutes ago
@killercowuk can you reproduce it at sandbox.propelorm.org?
killercowuk
22 minutes ago
Not tried, this, however having stepped through some more - I may have spotted part of the issue - I have not defined any of the fields as a primaryKey, is this known to be a requirement, do all tables defined in schema HAVE to have a primarykey for propel queries to work?
will try to recreate in sandbox just now though, thanks for the suggestion
marcj
20 minutes ago
well it’s highly recommended. why don’t you have a pk?
simple queries should work tho
killercowuk
16 minutes ago
The results are from a view that does some work grouping results etc. - this is a readonly, skipSQL table in the schema. I could artificially add a primarykey to it, but a primaryKey isn't really needed at present. If propel needs one, I could get SQL to invent one, but would rather not do this, from a cleanliness point of view [:blush:]
having stepped through the code, it appears the issue is because Propel is looking in the instancepool with a NULL key... which repeatedly returns the first result.
any suggestions to prevent this?
marcj
13 minutes ago
I see. well I believe we don’t have any VIEW releated test in our suite, so it’s actually not supported. although I don’t see a reason why it shouldn’t work.
ah ok, that makes probably sense. you could deactive instance pooling for this particular table
killercowuk
13 minutes ago
ah cool, how can I do this? (sorry still relatively new!)
I can't seem to find anything in database schema docs to allow pooling to be disabled..
marcj
9 minutes ago
you can disable it here http://api.propelorm.org/2.0-master/Propel/Runtime/Propel.html#method_disableInstancePooling
killercowuk
8 minutes ago
that would disable it for all queries to this connection then?
marcj
8 minutes ago
yes, globally. but you can enable it after your query again
temporarily workaround
killercowuk
3 minutes ago
that's done the trick, thanks marcj! Is this something worth raising a ticket for?
marcj
2 minutes ago
yes, would be cool
killercowuk
2 minutes ago
super will do, thanks for your all your help!
marcj
2 minutes ago
you’re welcome!
# 

So in short it seems Querying tables without PKs won't work when running with instance pools. The work around is to temporarily disable this functionality when querying:

Propel::disableInstancePooling();

Re-enable with:

Propel::enableInstancePooling();

Having stepped through the code I can see why this is an issue - and I'm not sure how you could get around this (perhaps create PKs based on hashing all column values together?! though this still doesn't guarantee uniqueness!). Even if there is no view to fix this, I hope anyone else bumping into this issue can be saved a few hours of head scratching.

**Comment by motin** [comment]
Any update on this? I experience the same symptoms, ie the first row is repeated. Mostly when using views. Maybe propel should throw an exception when instance-pooling is active in a scenario where it isn't supported - that way at least propel won't just silently return erroneous data.

**Comment by marcj** [comment]
In a ORM world working without a unique identifier per database row is the only way to know which php object is connected to which row in the database. Working without primary key makes ORM basically not useable as save-statements couldn't work anymore. So I'd rather go the way and forbid having a table definition without a primary key as it makes really no sense in the first place.

**Comment by marcj** [comment]
Yes, I agree.

**Comment by dereuromark** [comment]
@marcj Does https://github.com/spryker/support/issues/103 come from the same issue regarding the instance pooling?
Is there any way this could be fixed on propel side - or what do you suggest here to do.
We seem to have this issue in multiple places because of how propel internally handles this.
Thank you in advance for your feedback.

**Comment by marcj** [comment]
@dereuromark it seems your tables have primary keys, don't they?

**Comment by dereuromark** [comment]
Yes, those do. So it's a separate issue after all. Shall we open a different ticket then to not mix those up?
